File: bashpush.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"html"
	"io/ioutil"
	"log"
	"net/http"
	"os"
	"regexp"
	"strings"

	"github.com/SlyMarbo/rss"
	"github.com/sideshow/apns2"
	"github.com/sideshow/apns2/certificate"
	"github.com/sideshow/apns2/payload"
)

func debuglog(msg string) {
	if os.Getenv("BASHPUSH_DEBUG") == "1" {
		fmt.Println(msg)
	}
}

func main() {
	quote, link, id := getLastQuote()
	if isNew(id) {
		err := ioutil.WriteFile("lastquote.txt", []byte(id), 0644)
		if err == nil {
			// Panicking on the error but sending the notifications nonetheless
			// would spam pushover and slack every minute until I fix it due
			// to cron. Let's not do that.
			sendSlackNotification(id, link)
			sendiOSNotifications(id, quote)
		} else {
			debuglog(fmt.Sprintf("Couldn't write lastquote.txt, exiting. Err: %s", err))
		}
	} else {
		debuglog("Quote isn't new, exiting")
	}
}

func getLastQuote() (string, string, string) {

	bashFeedURL := os.Getenv("BASH_FEED_URL")
	if bashFeedURL == "" {
		panic("BASH_FEED_URL not set")
	}

	feed, err := rss.Fetch(bashFeedURL)
	if err != nil {
		panic(err)
	}

	r, _ := regexp.Compile(`\d+`)
	lastQuote := feed.Items[0]
	content := lastQuote.Content
	content = html.UnescapeString(content)

	debuglog(fmt.Sprintf("Found %s", lastQuote.Link))
	return content, lastQuote.Link, r.FindString(lastQuote.Link)
}

func isNew(id string) bool {
	if _, err := os.Stat("lastquote.txt"); os.IsNotExist(err) {
		// lastquote.txt does not yet exist
		err := ioutil.WriteFile("lastquote.txt", []byte{}, 0644)
		if err != nil {
			panic(err)
		}
	}
	last, err := ioutil.ReadFile("lastquote.txt")
	if err != nil {
		panic(err)
	}
	return id != strings.TrimSpace(string(last))
}

func sendiOSNotifications(id, quote string) {
	debuglog("Starting sending of iOS notifications")
	cert, err := certificate.FromP12File("./bashfsrPush.p12", "bashpush")
	if err != nil {
		log.Fatal(err)
	}

	devicesStr := os.Getenv("IOS_PUSH_TOKENS")
	devices := strings.Split(devicesStr, "\n")

	client := apns2.NewClient(cert).Development()

	for i := 0; i < len(devices); i++ {
		notification := &apns2.Notification{}
		notification.DeviceToken = devices[i]

		debuglog(fmt.Sprintf("Sending notification to %s", devices[i]))

		payload := payload.NewPayload()
		payload.AlertBody(quote)
		payload.Badge(1)
		payload.Sound("default")
		payload.Category("quote_category")
		payload.Custom("quote_id", id)
		notification.Payload = payload

		res, err := client.Push(notification)

		if err != nil {
			log.Fatal(err)
		}

		debuglog(fmt.Sprintf("%v %v %v\n", res.StatusCode, res.ApnsID, res.Reason))
	}
}

func sendSlackNotification(id, link string) {
	debuglog("Starting sending of Slack notification")
	hc := http.Client{}

	slackHookURL := os.Getenv("SLACK_HOOK_URL")
	if slackHookURL == "" {
		log.Fatal("SLACK_HOOK_URL not set")
	}

	quoteString := fmt.Sprintf("Neues Bash Zitat -> <%s|%s> 🤐", link, id)
	payload := map[string]string{"text": quoteString}
	jsonPayload, _ := json.Marshal(payload)

	req, _ := http.NewRequest("POST", slackHookURL, bytes.NewBufferString(string(jsonPayload)))
	resp, err := hc.Do(req)
	if err != nil {
		log.Fatal(err)
	}
	if resp.StatusCode != 200 {
		log.Fatal("Post to Slack failed /o\\")
	} else {
		debuglog("Finished sending of Slack notification without errors (I think)")
	}
}

File: bashpush.py
#! /usr/bin/env python

import os
from sys import exit, stderr
import re
import json
import requests
import feedparser
from pushover_complete import PushoverAPI


def env(var: str) -> str:
    """Read value from environment and exit if it doesn't exist."""
    val = os.getenv(var)
    if val is None:
        exit(f'missing env var {var}')
    return val


def fetch_latest() -> dict:
    """Fetch the newest quote from the RSS feed."""
    bash_url = env('BASH_FEED_URL')
    feed = requests.get(bash_url)
    # feedparser can't deal with the bash url directly 🙄
    latest = feedparser.parse(feed.text).entries[0]
    return {
        'url': latest['link'],
        'quote': latest['summary'],
        'id': int(re.findall(r'\?(\d+)', latest['link'])[0])
    }


def is_new(new_id: int) -> bool:
    """Check if the given quote id is newer than the last known one.
    If it is, it's persisted to `lastquote.txt`."""
    if not os.path.exists('lastquote.txt'):
        open('lastquote.txt', 'a').close()

    with open('lastquote.txt', 'r+') as f:
        last_id = f.read()
        last_id = 0 if last_id is '' else int(last_id)

        if new_id > last_id:
            f.seek(0)
            f.write(str(new_id))
            return True
        else:
            return False


def send_pushover_notifications(latest: dict):
    users = env('PUSHOVER_USER_TOKENS').split(',')
    p = PushoverAPI(env('PUSHOVER_API_TOKEN'))
    for user in users:
        p.send_message(user, latest['quote'], url=latest['url'])


def send_slack_notification(latest: dict):
    slack_url = env('SLACK_HOOK_URL')
    quote_str = f'Neues Bash Zitat 👉 <{latest["url"]}|{latest["id"]}> 🤐'
    quote_json = json.dumps({'text': quote_str})
    r = requests.post(slack_url, data=quote_json)
    if r.status_code is not 200:
        print(f'Post to Slack failed for {latest["id"]} /o\\\n{r.text}',
              file=stderr)


if __name__ == '__main__':
    latest = fetch_latest()
    if is_new(latest['id']):
        send_slack_notification(latest)
        send_pushover_notifications(latest)
